File: agent/plans/reviewer.py
from __future__ import annotations

import json
import re
from dataclasses import dataclass, field
from typing import Any

_VERDICT_BLOCK = re.compile(r"```verdict\s*(.*?)```", re.DOTALL | re.IGNORECASE)


@dataclass(frozen=True)
class ReviewerVerdict:
    passed: bool
    summary: str = ""
    commands: list[str] = field(default_factory=list)
    command_evidence: list[dict[str, Any]] = field(default_factory=list)

    def to_payload(self) -> dict[str, Any]:
        return {
            "passed": self.passed,
            "summary": self.summary,
            "commands": list(self.commands),
            "command_evidence": list(self.command_evidence),
        }


def parse_reviewer_verdict(text: str | None) -> ReviewerVerdict | None:
    if not text:
        return None
    blocks = _VERDICT_BLOCK.findall(text)
    if not blocks:
        return None
    raw = blocks[-1].strip()  # last block wins
    try:
        data = json.loads(raw)
    except json.JSONDecodeError:
        return None
    if not isinstance(data, dict) or "passed" not in data:
        return None
    commands = [str(item) for item in (data.get("commands") or []) if str(item).strip()]
    evidence = [item for item in (data.get("command_evidence") or []) if isinstance(item, dict)]
    return ReviewerVerdict(
        passed=bool(data.get("passed")),
        summary=str(data.get("summary") or "")[:1000],
        commands=commands,
        command_evidence=evidence,
    )

File: packages/core/src/plans/reviewer.ts
/**
 * 独立 reviewer verdict 解析 (MIG-CTRL-008)。对齐 Python `agent/plans/reviewer.py`。
 * 解析 ```verdict ...``` JSON 块（最后一个生效）。
 */
const VERDICT_BLOCK = /```verdict\s*([\s\S]*?)```/gi

export interface ReviewerVerdict {
  passed: boolean
  summary: string
  commands: string[]
  commandEvidence: Array<Record<string, unknown>>
}

export function verdictToPayload(v: ReviewerVerdict): Record<string, unknown> {
  return {
    passed: v.passed,
    summary: v.summary,
    commands: [...v.commands],
    command_evidence: [...v.commandEvidence],
  }
}

export function parseReviewerVerdict(
  text: string | null | undefined,
): ReviewerVerdict | null {
  if (!text) return null
  const blocks = [...text.matchAll(VERDICT_BLOCK)].map((m) => m[1]!)
  if (!blocks.length) return null
  const raw = blocks[blocks.length - 1]!.trim() // last block wins
  let data: unknown
  try {
    data = JSON.parse(raw)
  } catch {
    return null
  }
  if (
    !data ||
    typeof data !== 'object' ||
    Array.isArray(data) ||
    !('passed' in data)
  )
    return null
  const obj = data as Record<string, unknown>
  const commands = ((obj.commands ?? []) as unknown[])
    .map((item) => String(item))
    .filter((item) => item.trim())
  const evidence = ((obj.command_evidence ?? []) as unknown[]).filter(
    (item) => item && typeof item === 'object',
  ) as Array<Record<string, unknown>>
  return {
    passed: Boolean(obj.passed),
    summary: String(obj.summary ?? '').slice(0, 1000),
    commands,
    commandEvidence: evidence,
  }
}
